src/transaction/log_manager.c

```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+	      }
+
+	    break;
+	  }
+	case LOG_UNDO_DATA:
+	  {
+	    LOG_REC_UNDO *undo = NULL;
+
+	    undo = (LOG_REC_UNDO *) (log_page_p->area + process_lsa.offset);
+	    rcvindex = undo->data.rcvindex;
+	    undo_length = undo->length;
+
+	    if (rcvindex == RVOVF_PAGE_UPDATE)
+	      {
+		/* GET OVF UNDO IMAGE */
+		if ((error_code =
+		     cdc_get_overflow_recdes (thread_p, log_page_p, undo_recdes, *undo_lsa, rcvindex,
+					      false)) != NO_ERROR)
+		  {
+		    goto end;
+		  }
+
+	      }
+	    else if (rcvindex == RVHF_MVCC_UPDATE_OVERFLOW)
+	      {
+		LOG_READ_ADD_ALIGN (thread_p, sizeof (*undo), &process_lsa, log_page_p);
+
+		if (ZIP_CHECK (undo_length))
+		  {
+		    undo_length = (int) GET_ZIP_LEN (undo_length);
+		    is_zipped_undo = true;
+		  }
+
+		if (process_lsa.offset + undo_length < (int) LOGAREA_SIZE)
+		  {
+		    undo_data = (char *) (log_page_p->area + process_lsa.offset);
+		  }
+		else
+		  {
+		    undo_data = (char *) malloc (undo_length);
+		    if (undo_data == NULL)
+		      {
+			error_code = ER_OUT_OF_VIRTUAL_MEMORY;
+			goto end;
+		      }
+
+		    logpb_copy_from_log (thread_p, undo_data, undo_length, &process_lsa, log_page_p);
+		    is_undo_alloced = true;
+		  }
+
+		if (is_zipped_undo && undo_length != 0)
+		  {
+		    undo_zip_ptr = log_zip_alloc (IO_PAGESIZE);
+		    if (undo_zip_ptr == NULL)
+		      {
+			error_code = ER_OUT_OF_VIRTUAL_MEMORY;
+			goto end;
+		      }
+
+		    is_unzipped_undo = log_unzip (undo_zip_ptr, undo_length, undo_data);
+		    if (is_unzipped_undo != true)
+		      {
+			error_code = ER_IO_LZ4_DECOMPRESS_FAIL;
+			goto end;
+		      }
+		  }
+
+		if (is_zipped_undo && is_unzipped_undo)
+		  {
+		    undo_length = (int) undo_zip_ptr->data_length;
+		    if (is_undo_alloced)
+		      {
+			free_and_init (undo_data);
+		      }
+		    undo_data = undo_zip_ptr->log_data;
+		  }
+
+		undo_recdes->type = *(INT16 *) undo_data;
+		undo_recdes->length = undo_length - sizeof (INT16);
+		undo_recdes->data = (char *) malloc (undo_recdes->length);
+		if (undo_recdes->data == NULL)
+		  {
+		    error_code = ER_OUT_OF_VIRTUAL_MEMORY;
+		    goto end;
+		  }
+
+		memcpy (undo_recdes->data, (char *) undo_data + sizeof (undo_recdes->type), undo_recdes->length);
+	      }
+
+	    break;
+	  }
+
+	case LOG_REDO_DATA:
+	  {
+	    LOG_REC_REDO *redo = NULL;
+	    LOG_RCVINDEX rcvindex;
+
+	    redo = (LOG_REC_REDO *) (log_page_p->area + process_lsa.offset);
+	    rcvindex = redo->data.rcvindex;
+
+	    if (rcvindex == RVOVF_PAGE_UPDATE)
+	      {
+		/* GET OVF UNDO IMAGE */
+		if ((error_code =
+		     cdc_get_overflow_recdes (thread_p, log_page_p, undo_recdes, *undo_lsa, rcvindex,
+					      false)) != NO_ERROR)
+		  {
+		    goto end;
+		  }
+	      }
+
+	    break;
+	  }
+
+	default:
+	  break;
+	}
+
+      if (is_undo_alloced)
+	{
+	  free_and_init (undo_data);
+	  is_undo_alloced = false;
+	}
+    }
+
+/* Get REDO RECDES from redo lsa */
+  if (redo_lsa != NULL)
```

Comment:
Have you considered the generalized way to extract a redo log record like log_recovery_redo()? They don't deal with log records rcvindex by rcvindex. I guess you see that code already. please tell me why you should do it case by case?